src/utils.jl

```diff
@@ -29,3 +29,69 @@ function compute_discontinuity_tree(lags, alg, start_val)
     start_val + unique(vcat((sum.(collect(with_replacement_combinations(lags, i))) for i in
                              1:alg_order(alg))...))
 end
+
+"""
+    reduce_solution!(integrator::DDEIntegrator, tmax)
+
+Remove time points of ODE solution of `integrator` up to time `tmax` that are not required
+for calculation of DDE solution.
+"""
+@inline function reduce_solution!(integrator::DDEIntegrator, tmax)
+    if !integrator.opts.dense && !integrator.opts.save_everystep &&
+        integrator.saveiter < integrator.integrator.saveiter
+
+        # last time point of ODE solution that is known to be required for DDE solution
+        @inbounds t_sol_prev = integrator.integrator.sol.t[integrator.saveiter]
+        # next time point of ODE solution that might be required for DDE solution
+        @inbounds t_sol = integrator.integrator.sol.t[integrator.saveiter + 1]
+
+        @inbounds while t_sol < tmax && integrator.saveiter < integrator.integrator.saveiter
+            needed = false
+
+            while !(isempty(integrator.saveat)) && top(integrator.saveat) <= t_sol
+                # do not remove time point if it is upper bound of a required
+                # interpolation interval
+                if t_sol_prev < top(integrator.saveat)
+                    needed = true
+                end
+
+                # remove passed time points
+                pop!(integrator.saveat)
+            end
+
+            # do not remove time point if it is lower bound of a required interpolation
+            # interval
+            needed = needed || (!(isempty(integrator.saveat)) &&
+                                integrator.saveiter + 1 < integrator.integrator.saveiter &&
+                                t_sol < top(integrator.saveat) <
+                                integrator.integrator.sol.t[integrator.saveiter + 2])
+
+            if !needed
+                # delete not required time point, function value, and interpolation data
+                deleteat!(integrator.integrator.sol.t, integrator.saveiter + 1)
```

Comment:
I'd run timings on it. I know Julia amortizes `push!`, but I think it might re-allocate with eery `deleteat!` and subsequently ignore the amortization, making it quite expensive to `delateat!` very often. I would only want to use this code path then in cases where I have a really tight RAM limit that I am hitting. It would probably be good to keep it as an option.